### ISSUE TITLE
new lint: sort_child_properties_last

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#1542]
Name TBD but the idea is to nudge people to sorting child/children properties last in their build methods to improve UI-as-Code feature readability.

@jacob314: could you add some more context?  Maybe some concrete exemplars?

**Comment by a14n** [comment]
Close to #1129.
Alternatively we could add a lint to ensure named arguments are provided in the same order as in the method declaration.

**Comment by jacob314** [comment]
We should do both. Inside package:flutter it would make sense to lint that named arguments that are `Widget` and `Iterable<Widget>`  occur last in the list of arguments.
From my perspective I don't care if named arguments are out of order as long as they are sorted so Widget and `Iterable<Widget>` arguments are last.

**Comment by pq** [comment]
@jacob314: could you add some real-world examples and help me fill in the motivation?

Spitballing a description:

```
Sort child properties last in widget instance creations.  This improves
readability and generally plays nicest with auto-formatting and UI-as-code 
language constructs.
```

**Comment by jacob314** [comment]
Here is an example of how violating this pattern hurts the utility of the UI as Code Guides.
### Bad property sorting.
Notice it is hard to see what widget the `onPressed`, and `mainAxisAlignment` properties are associated with as they occur after the lines are drawn.  Properties in the correct order appear clearly associated with the constructor call and separated from the children.
<img width="548" alt="Screen Shot 2019-04-28 at 11 04 53 AM" src="https://user-images.githubusercontent.com/1226812/56868312-8e558380-69a5-11e9-9434-8eb9ff5dfe42.png">

### With properties listed before child widgets:
<img width="555" alt="Screen Shot 2019-04-28 at 11 04 16 AM" src="https://user-images.githubusercontent.com/1226812/56868352-2bb0b780-69a6-11e9-9bb9-17074f46092a.png">

Background info:
In the inspector tree view, and `toStringDeep` text dumps of Widget trees, children are always listed last. This lint brings lints for how code authors write code in alignment. UI as code widget tree visualization makes widget build methods look more similar to inspector and toStringDeep dumps for trees so having alignment now has more value than previously.

Most constructor definitions and constructor calls in package:flutter already match this lint so this lint is for the most part matching the facts on the ground and enforcing consistency instead of changing how most code in the framework is written.

Notice that child and children are listed last in these constructor definitions:
```dart
  Row({
    Key key,
    MainAxisAlignment mainAxisAlignment = MainAxisAlignment.start,
    MainAxisSize mainAxisSize = MainAxisSize.max,
    CrossAxisAlignment crossAxisAlignment = CrossAxisAlignment.center,
    TextDirection textDirection,
    VerticalDirection verticalDirection = VerticalDirection.down,
    TextBaseline textBaseline,
    List<Widget> children = const <Widget>[],
  })

  ConstrainedBox({
    Key key,
    @required this.constraints,
    Widget child,
  })  : assert(constraints != null),
        assert(constraints.debugAssertIsValid()),
        super(key: key, child: child);
```

**Comment by a14n** [comment]
Perhaps the lint should apply only if there's only one argument that is Widget or Iterable<Widget>.

For instance, putting Widgets at the end of argument list will dissociate the widget from its alignment for the following sample:

```dart
MyWidget({
  CrossAxisAlignment headerAlignment = CrossAxisAlignment.center,
  Widget header,
  CrossAxisAlignment contentAlignment = CrossAxisAlignment.start,
  Widget content,
})
```

**Comment by pq** [comment]
Good example @a14n.

TBH: my initial instinct is to be more narrow still and limit it to `child` and `children` properties to avoid possible (and potentially confusing / annoying) false positives.

**Comment by pq** [comment]
@jacob314: the first cut has landed.  Feel free to chime in here or separately if you'd like to see any tweaks to description and/or content. 👍

**Comment by a14n** [comment]
I'm trying to apply this new lint to flutter code base and I find some case less readable IMH (especially when callback is used). For instance:

```dart
//-------------- before
RaisedButton(
  child: const Text('CONFIRMATION'),
  onPressed: () {
    showTimePicker(
      context: context,
      initialTime: _selectedTime,
    )
    .then<void>((TimeOfDay value) {
      if (value != null && value != _selectedTime) {
        _selectedTime = value;
        _scaffoldKey.currentState.showSnackBar(SnackBar(
          content: Text('You selected: ${value.format(context)}'),
        ));
      }
    });
  },

//-------------- after
RaisedButton(
  onPressed: () {
    showTimePicker(
      context: context,
      initialTime: _selectedTime,
    )
    .then<void>((TimeOfDay value) {
      if (value != null && value != _selectedTime) {
        _selectedTime = value;
        _scaffoldKey.currentState.showSnackBar(SnackBar(
          content: Text('You selected: ${value.format(context)}'),
        ));
      }
    });
  },
  child: const Text('CONFIRMATION'),
),
```


WDYT?

**Comment by pq** [comment]
Thanks for doing this!

My understanding is that this formatting really shines w/ the method guides being added to IDEs...  I'd be curious though if there are cases where we don't want to recommend it?

/cc @jacob314

**Comment by a14n** [comment]
@jacob314 @hixie regarding my previous comment, does the result look good too you?

**Comment by jacob314** [comment]
Here are screenshots of that example with child at the start of the list and at the end of the list. In this case the child widget tree is trivial so the difference isn't huge. This convention matters more when the child widget tree is larger so it is easier to lose your place in the code.
I've added a second slightly more complex example of a real world use case ofRaisedButton from the Flutter Gallery. Notice it is a lot easier to immediately spot what widget the properties are associated with.
Before:
<img width="578" alt="Screen Shot 2019-05-08 at 10 30 32 AM" src="https://user-images.githubusercontent.com/1226812/57395202-59eb7f80-717c-11e9-86ee-8f1166088aef.png">
After:
<img width="593" alt="Screen Shot 2019-05-08 at 10 30 18 AM" src="https://user-images.githubusercontent.com/1226812/57395205-5c4dd980-717c-11e9-91b3-64b6a8d1cf94.png">

Complex example -- before:
<img width="480" alt="Screen Shot 2019-05-08 at 10 34 53 AM" src="https://user-images.githubusercontent.com/1226812/57395530-2fe68d00-717d-11e9-8097-b0955bb64f9c.png">
<img width="498" alt="Screen Shot 2019-05-08 at 10 34 26 AM" src="https://user-images.githubusercontent.com/1226812/57395528-2d843300-717d-11e9-8949-c77002f54f22.png">

**Comment by jacob314** [comment]
> ```dart
> MyWidget({
>   CrossAxisAlignment headerAlignment = CrossAxisAlignment.center,
>   Widget header,
>   CrossAxisAlignment contentAlignment = CrossAxisAlignment.start,
>   Widget content,
> })
> ```

Interesting example @a14n.
Interleaving properties with widget children is fine as long as the properties are always before the associated widget. In that case it could be reasonable to require that the named arguments are specified in the order they are specified in the `MyWidget` constructor which would keep the properties right next to the associated widget.
Bad example showing properties after the relevant widget.
<img width="383" alt="Screen Shot 2019-05-08 at 10 47 12 AM" src="https://user-images.githubusercontent.com/1226812/57396185-acc63680-717e-11e9-8aa8-7ad304e9de78.png">
Ok examples showing properties before children.
<img width="452" alt="Screen Shot 2019-05-08 at 10 46 51 AM" src="https://user-images.githubusercontent.com/1226812/57396186-acc63680-717e-11e9-9141-4b5d90e26b84.png">
Ok example showing properties interleaved:
<img width="462" alt="Screen Shot 2019-05-08 at 10 46 23 AM" src="https://user-images.githubusercontent.com/1226812/57396187-acc63680-717e-11e9-9725-1d6ef180607b.png">

**Comment by a14n** [comment]
With the _UI Guides_ (that I don't have for now on VSCode) my complaint on the sample with _onPressed_ almost disappear. (I didn't understand what was those stronger lines in your screenshot until I read https://medium.com/dartlang/announcing-dart-2-3-optimized-for-building-user-interfaces-e84919ca1dff)

**Comment by jacob314** [comment]
Let me know if you would like to beta-test UI Guides in VSCode when we get them working.

**Comment by Hixie** [comment]
I think this definitely helps in many cases but realistically I think there'd be too many false positives. At the end of the day, at some level, we have to let the programmer figure out what is best for them.

**Comment by jacob314** [comment]
This lint is now working well for google projects.
We might want to look at creating a weakened version that is appropriate to turn on for all projects. The weakened version would allow arguments where the value is a Function to show up after `child` properties. As those arguments can be very large expressions, placing them first is less of a clear win particularly if the widget tree is simple.

**Comment by Yetispapa** [comment]
Is there a lint rule, to have the complete opposite, so that the child_properties are first?

**Comment by bwilkerson** [comment]
No, there isn't. Is there a motivation for such a rule?

**Comment by Yetispapa** [comment]
I learned this order back then and I apply it ever since then. The order for me is:

```
private properties
public properties
constructor
public functions
private functions
```

or

```
public properties
constructor
public functions
private properties
private functions
```

I think, back then in most schools / universities the order was always properties first and then functions and that's where i'm coming from.

Was just asking, maybe I will used to the new one :)

**Comment by bwilkerson** [comment]
I believe that there is a misunderstanding. The [sort_child_properties_last](https://dart.dev/tools/linter-rules/sort_child_properties_last) lint enforces an order for named arguments in the invocation of constructors of Flutter `Widget` subclasses. It doesn't enforce an ordering of members in a class.

There is a 'Sort Members' command in the IDEs that will sort the members in classes, as well as top-level declarations, in a predefined order, but there is no lint to enforce that order.
